pkg/controllers/node/controller.go

```diff
@@ -45,50 +51,84 @@ func NewController(kubeClient client.Client) *Controller {
-	// 1. Retrieve node from reconcile request
+	// 1. Retrieve Node, ignore if not provisioned or terminating
 	stored := &v1.Node{}
 	if err := c.kubeClient.Get(ctx, req.NamespacedName, stored); err != nil {
 		if errors.IsNotFound(err) {
 			return reconcile.Result{}, nil
 		}
-		return reconcile.Result{}, err
+		return result.RetryIfError(ctx, err)
 	}
 	if _, ok := stored.Labels[v1alpha3.ProvisionerNameLabelKey]; !ok {
 		return reconcile.Result{}, nil
 	}
+	if !stored.DeletionTimestamp.IsZero() {
+		return reconcile.Result{}, nil
+	}
 
-	// 2. Execute node reconcilers
+	// 2. Retrieve Provisioner
+	provisioner := &v1alpha3.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: stored.Labels[v1alpha3.ProvisionerNameLabelKey]}, provisioner); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// 3. Execute reconcilers
 	node := stored.DeepCopy()
+	var backoffs []time.Duration
 	var errs error
 	for _, reconciler := range []interface {
-		Reconcile(*v1.Node) error
+		Reconcile(context.Context, *v1alpha3.Provisioner, *v1.Node) (reconcile.Result, error)
 	}{
 		c.readiness,
+		c.expiration,
 		c.finalizer,
 	} {
-		errs = multierr.Append(errs, reconciler.Reconcile(node))
+		res, err := reconciler.Reconcile(ctx, provisioner, node)
+		errs = multierr.Append(errs, err)
+		backoffs = append(backoffs, res.RequeueAfter)
 	}
 
-	// 3. Patch any changes, regardless of errors
+	// 4. Patch any changes, regardless of errors
 	if !reflect.DeepEqual(node, stored) {
 		if err := c.kubeClient.Patch(ctx, node, client.MergeFrom(stored)); err != nil {
 			return reconcile.Result{}, fmt.Errorf("patching node %s, %w", node.Name, err)
 		}
 	}
+
+	// 5. Requeue if error or if retryAfter is set
+	if requeueAfter := functional.MaxDuration(backoffs...); requeueAfter != 0 {
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
+	}
 	return result.RetryIfError(ctx, errs)
 }
 
-func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
 	return controllerruntime.
 		NewControllerManagedBy(m).
 		Named("Node").
 		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha3.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				nodes := &v1.NodeList{}
+				if err := c.kubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha3.ProvisionerNameLabelKey: o.GetName()})); err != nil {
+					logging.FromContext(ctx).Errorf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			}),
+		).
 		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
```

Comment:
I realize this was included in a prior PR, but why did we pick 10 here? Are there any detriments to bumping this number up?

Reply:
We could consider setting this to kubeclient qps (default 20). I found that 10 was reasonably high parallelism without exhausting default qps. Perhaps we can evaluate this number as we do more scale testing.